### PR TITLE
import from .mat-files

### PR DESCRIPTION
**Pull request by ALEhmerDLR**
Update changed values from .mat-files which are generated by VirSat or look like this

can Update:

Value of ValuePropertyInstance
Value and Unit of UnitValuePropertyInstance
Value and Unit of EnumUnitPropertyInstance
Value of ResourcePropertyInstance
Value of ReferencePropertyInstance
Value of EReferencePropertyInstance

Content of Composed Category if it is the same Category as before export
Content of each element of ArrayProperty if they not changed their Type

Closes #525

### CODE REVIEW

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+		exporter = new MatExporter();
+		importer = new MatImporter();
+		super.setUp();
+		addEditingDomainAndRepository();
+		executeAsCommand(() -> loadTestConcept());
+		tsei = new TestStructuralElement(testConcept);
+		sei = tsei.getStructuralElementInstance();
+		sei.setName("testsei");
+		executeAsCommand(() -> repository.getRootEntities().add(sei));
+		mat = exporter.exportSei(sei);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Could we also have one test that doesn't use the exporter to test the importer? By creating this matlab-string manually for example? Because like this, if the exporter might have a bug, we would maybe also miss bugs in the importer test

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+				sei.getCategoryAssignments().get(0).getName().equals(tc2.getName()));
+	}
+
+	@Test
+	public void testImportAPIDeleteAPI() throws IOException {
+		EReferenceTest tc2 = new EReferenceTest(testConcept);
+		tsei.add(tc2);
+		TestCategoryAllProperty tc3 = new TestCategoryAllProperty(testConcept);
+		tsei.add(tc3);
+		APropertyInstance nInstance = sei.getCategoryAssignments().get(0).getPropertyInstances().get(0);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: It's reletively hard to follow this test as someone from the outside... All this index handling and the connection of tsei and sei are hard to understand... Could you add som comments in your tests?

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+		tc1.setTestResource(testUri);
+		sei2.setName("testsei");
+		sei2.setUuid(sei.getUuid());
+		tsei2.add(tc1);
+		importer.importSei(sei2, mat);
+
+		EList<APropertyInstance> caSei = sei.getCategoryAssignments().get(0).getPropertyInstances();
+		EList<APropertyInstance> caSei2 = sei2.getCategoryAssignments().get(0).getPropertyInstances();
+		assertTrue("same number of elements", caSei.size() == caSei2.size());
+		assertEquals("same testString", tc1.getTestString(), tc.getTestString());
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: You mixed up expected and actual here, didn't you? You expect that all values of tc1 are overwritten by the import of tc? Then it the order in the assert statement should be the other way around... This way it is confusing because you think that tc should have change to the values you set to tc1.... And please also add a comment here or a more meaningful name of the test case

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+		assertEquals("same testBool", tc1.getTestBool(), tc.getTestBool());
+		assertEquals(tc1.getTestFloat(), tc.getTestFloat(), 0);
+		assertEquals("same testResource", tc1.getTestResource(), tc.getTestResource());
+		assertEquals("same testEnum", tc1.getTestEnum(), tc.getTestEnum());
+		
+	}
+
+	@Test
+	public void testImportOfValuesAddAll() throws IOException {
+		TestCategoryAllProperty tc = new TestCategoryAllProperty(testConcept);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Could you create these values as contants and then assert against these constants? As it is write now it might happen that tc changes for some reason and then the assert statements are invalid..
Something like:
final String TEST_STRING = "TestString";
tc.setTestString(TEST_STRING);
...
assert("Some comment", TEST_STRING, tc1.getTestString());

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		assertTrue("same number of elements", caSei.size() == caSei2.size());
+		assertEquals("same testString", tc1.getTestString(), tc.getTestString());
+		assertEquals("same testBool", tc1.getTestBool(), tc.getTestBool());
+		assertEquals(tc1.getTestFloat(), tc.getTestFloat(), 0);
+		assertEquals("same testResource", tc1.getTestResource(), tc.getTestResource());
+		assertEquals("same testEnum", tc1.getTestEnum(), tc.getTestEnum());
+	}
+
+	@Test
+	public void testImportOfValuesChangeNothingValues() throws IOException {
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: ChangeNothingValues? TestCase with no updates?

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+		assertEquals("same testResource", tc1.getTestResource(), tc.getTestResource());
+		assertEquals("same testEnum", tc1.getTestEnum(), tc.getTestEnum());
+		assertEquals("same testInt", tc1.getTestInt(), tc.getTestInt());
+	}
+
+	@Test
+	public void testImportOfValuesRef() throws IOException {
+		//empty and import empty
+		TestCategoryReference tc = new TestCategoryReference(testConcept);
+		tsei.add(editingDomain, tc);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: This method just returns you a command... which is not executed as long as you don't invoke its execution... 
Command cmd = tsei.add(edetingDomain, tc);
edetingDomain.execute(cmd);

Comment (franzTobiasDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Looks like this test isn't doing anything till now, is it? If values are not set then this CA is still empty

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+		mat = exporter.exportSei(sei);
+		MatFile mat1 = exporter.exportSei(sei);
+		TestStructuralElement tsei2 = new TestStructuralElement(testConcept);
+		StructuralElementInstance sei2 = tsei2.getStructuralElementInstance();
+		EReferenceTest tc1 = new EReferenceTest(testConcept);
+		tsei2.add(editingDomain, tc1);
+		sei2.setName("testsei");
+		sei2.setUuid(sei.getUuid());
+		importer.importSei(sei2, mat);
+		assertEquals("same Reference from empty to empty", tc1.getEReferenceTest(), tc.getEReferenceTest());
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: I think it makes more sense to test the value of an EReference here... rather than comparing the property...

File: de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+					importGivenAPI(seiAPIs.get(i), struct);
+				}
+				nameMatAPIs.remove(seiAPIs.get(i).getType().getName());
+				i++;
+			} else {
+				seiAPIs.remove(seiAPIs.get(i));
+			}
+		}
+
+		//import all new APropertyInstances
```
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java]: It is deleted now.

File: de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+	 */
+	private boolean checkIfCorrectSei(StructuralElementInstance sei, Struct seiStruct) {
+		EList<StructuralElementInstance> seiChildren = sei.getChildren();
+		List<String> structFields = seiStruct.getFieldNames();
+
+		//check Type and Uuid
+		if (structFields.contains("type") && structFields.contains("uuid")) {
+			String matSeiType = shorter(seiStruct.getChar("type").toString()); 
+			String matSeiUUID = shorter(seiStruct.getChar("uuid").toString());
+			structFields.remove(0);
```
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java]: I can remove them explicity. I changed this.

File: de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+	 * import a given EnumUnitPropertyInstance
+	 * 
+	 * updates name and unit. value is updates automatically
+	 * @param element PropertyInstance which should be changed
+	 * @param struct MatStruct that includes all Information
+	 */
+	private Boolean contentOfProperty(EnumUnitPropertyInstance element, Struct struct) {
+		if (element.getType() instanceof EnumProperty) {
+			BeanPropertyEnum bpe = new BeanPropertyEnum(element);
+			bpe.setValue(shorter(struct.get(MatHelper.NAME).toString()));
```
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java]: It sets the Value null.

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 5 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: A TestCase with no Values is exported and than imported. The test check if the sei doesn't change by this.

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 4 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Changed it.

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 3 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: I changed the order and added some comments.

File: de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java]: If there was a value set and you want to import a new value which doesn't correspond to any Enum Literal, it will set the value to null.

File: de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+	 * @param element PropertyInstance which should be changed
+	 * @param struct MatStruct that includes all Information
+	 */
+	private Boolean contentOfProperty(ResourcePropertyInstance element, Struct struct) {
+		if (element.getType() instanceof ResourceProperty) {
+			BeanPropertyResource bpr = new BeanPropertyResource(element);
+			if ("''".equals(struct.get(MatHelper.URI).toString())) {
+				bpr.unset();
+			} else {
+				bpr.setValue(URI.createURI(shorter(struct.get(MatHelper.URI).toString()), true));
```
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporter.java]: It is.

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+		assertFalse("Not the same type", importer.checkIfCorrectSei(sei2, mat));
+	}
+
+	@Test
+	public void testCheckIfCorrectSeiWrongNumberOfChildren() {
+		TestStructuralElementOther tsei2 = new TestStructuralElementOther(testConcept);
+		StructuralElementInstance sei2 = tsei2.getStructuralElementInstance();
+		sei2.setName("testsei");
+		sei2.setParent(sei);
+		assertFalse("Not the same number of children", importer.checkIfCorrectSei(sei, mat));
```
[the remark this answers is not in the document]
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: I changed it that way it is checking UUID and Type. Now children can be added and it will get the sei as a fitting one. Correct means that the sei and the one in the .mat fit to each other.

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 1 above)
Comment (franzTobiasDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Nice 👍 Thanks :)

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+		assertEquals("Is the same", TEST_BOOL, tc.getTestBool());
+		assertEquals("Is the same", TEST_INT, tc.getTestInt());
+		assertEquals(TEST_FLOAT, tc.getTestFloat(), 0);
+		assertEquals("Is the same", null, tc.getTestResource());
+	}
+	
+	@Test
+	public void testCheckIfCorrectSeiCorrect() {
+		assertTrue("Is the same", importer.checkIfCorrectSei(sei, mat));
+		sei.setName("Testsein");
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: So we prevent renaming of SEIs? Why? Aren't UUIDs itendfier enough?

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+		TestStructuralElementOther tsei2 = new TestStructuralElementOther(testConcept);
+		StructuralElementInstance sei2 = tsei2.getStructuralElementInstance();
+		sei2.setName(TEST_SEI);
+		sei2.setType(sei.getType());
+		assertFalse("Not the same uuid", importer.checkIfCorrectSei(sei2, mat));
+	}
+
+	@Test
+	public void testCheckIfCorrectSeiWrongType() {
+		TestStructuralElementOther tsei2 = new TestStructuralElementOther(testConcept);
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: That test is legacy or? It doesn't do anything else than the upper test? Or am I missing something?

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+		sei2.setName(TEST_SEI);
+		sei2.setUuid(sei.getUuid());
+		assertFalse("Not the same type", importer.checkIfCorrectSei(sei2, mat));
+	}
+
+	@Test
+	public void testImportUnit() throws IOException {
+		TestCategoryAllProperty tc = new TestCategoryAllProperty(testConcept);
+		tsei.add(tc);
+		tc.getTestFloatBean().setUnit("Kilogram");
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Would be nice if the Unit would be in a constant as well... :) 
`final String UNIT_NAME = "Kilogram";`

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 16]
```diff
@@ -0,0 +1,464 @@
...
+		mat = exporter.exportSei(sei);
+		TestStructuralElement tsei2 = new TestStructuralElement(testConcept);
+		StructuralElementInstance sei2 = tsei2.getStructuralElementInstance();
+		sei2.setName(TEST_SEI);
+		sei2.setUuid(sei.getUuid());
+		TestCategoryAllProperty tc1 = new TestCategoryAllProperty(testConcept);
+		tsei2.add(tc1);
+		editingDomain.getCommandStack().execute(importer.importSei(editingDomain, sei2, mat));
+		assertEquals("Unit set", "Kilogram", tc1.getTestFloatBean().getUnit());
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]:
```suggestion
		assertEquals("Unit set", UNIT_NAME, tc1.getTestFloatBean().getUnit());

```

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java [hunk 17]
```diff
@@ -0,0 +1,464 @@
...
+		assertEquals("same testEnumValue", tc.getTestEnumBean().getUnit(), tc1.getTestEnumBean().getUnit());
+		assertEquals("same testInt", TEST_INT, tc1.getTestInt());
+	}
+
+	@Test
+	public void testImportOfValuesRef() throws IOException {
+		//Reference test target
+		final TestCategoryAllProperty TEST_REFERENCE_TARGET = new TestCategoryAllProperty(testConcept);
+		
+		//empty and import empty
```
Comment (franzTobiasDLR) [inline de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Missleading comment... Test is further down. Rather "Prepare test data" or remove it

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 14 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: One checks for UUID, one for TYPE

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 13 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Right now yes. The name is used to take the right Entry of the .mat. So if it is the wrong name, it will not work yet

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 17 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Done

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 16 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Done

File: de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java (hunk 15 above)
Comment (ALEhmerDLR) [reply de.dlr.sc.virsat.model.extension.tests.test/src/de/dlr/sc/virsat/model/dvlm/mat/MatImporterTest.java]: Done
